### PR TITLE
Adding support for parse URI for hostnames

### LINKED ISSUE #1500: Implement HOST url parsing as part of parse_url support
HOST parsing will consist of pulling out the host portion of the uri, such as `www.google.com` in `https://www.google.com/q/cats`. It will need to conform to the spark interpretation of what constitutes a hostname and not the RFC as there are discrepancies.

### PR DESCRIPTION
**Pull request by hyperbolic2346**
**This PR is branched off of #1554 and a vast majority of the code is from that PR. Once that merges, this should reduce down to very little change.**

This PR adds support for hostname parsing in parse uri. This adds the tests and exposes the functionality.

closes #1500

### CODE REVIEW

**Comment by thirtiseven** [comment]
Few mismatches from integration test cases:

| url | cpu | gpu |
|---|---|---|
| http://userinfo@spark.apache.org/path?query=1#Ref | spark.apache.org | null |
| https://use%20r:pas%20s@example.com/dir%20/pa%20th.HTML?query=x%20y&q2=2#Ref%20two | example.com | example |
| http://userid:password@example.com:8080 | example.com | null |
| http://userid:password@example.com:8080/ | example.com | null |
| http://userid:password@example.com | example.com | exampl |
| http://userid:password@example.com/ | example.com | exampl |
| http://-.~_!$&'()*+,;=:%40:80%2f::::::@example.com | example.com | null |
| http://. | null | . |
| http://.. | null | .. |
| http://../ | null | .. |
| http://.www.foo.bar/ | null | .www.foo.bar |
| http://.www.foo.bar./ | null | .www.foo.bar. |

- URLs containing USERINFO are not handled correctly.
- hostnames have their own validation logic, they're generated by following syntax:
```
// hostname      = domainlabel [ "." ] | 1*( domainlabel "." ) toplabel [ "." ]
// domainlabel   = alphanum | alphanum *( alphanum | "-" ) alphanum
// toplabel      = alpha | alpha *( alphanum | "-" ) alphanum
```

**Comment by hyperbolic2346** [comment]
build

**Comment by thirtiseven** [comment]
All tests from plugin side passed (IT and the kaggle dataset), I think it's functional correct.

File: src/main/cpp/tests/parse_uri.cpp [hunk 1]
```diff
@@ -22,141 +22,175 @@
...
-     "http://www.nvidia.com/tags.php?%2F88\323\351\300ึณวน\331\315\370%2F",
-     "http://www.nvidia.com//wp-admin/includes/index.html#9389#123",
-     "http://www.nvidia.com/"
-     "object.php?object=ะก-\320%9Fะฑ-ะฟ-ะก\321%82\321%80ะตะป\321%8Cะฝะฐ-\321%83ะป-\320%"
-     "97ะฐะฒะพะด\321%81ะบะฐ\321%8F.html&sid=5",
-     "http://www.nvidia.com/picshow.asp?id=106&mnid=5080&classname=\271\253ืฐฦช",
-     "http://-.~_!$&'()*+,;=:%40:80%2f::::::@example.com:443",
-     "http://userid:password@example.com:8080/"});
-
+  auto col    = get_test_data(test_types::SPARK_EDGES);
```
Comment (ttnghia) [inline src/main/cpp/tests/parse_uri.cpp:144]: Nit: `auto const` for all variables.

**Review by ttnghia — DISMISSED** [review]

**Comment by hyperbolic2346** [comment]
build

**Review by ttnghia — APPROVED** [review]

**Pull request merged**